controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java

```diff
@@ -0,0 +1,104 @@
+
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final ScheduledExecutorService executor;
+
+    public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
+                                 final StreamMetadataStore streamMetaStore,
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetaStore);
+        Preconditions.checkNotNull(streamMetaTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataStore = streamMetaStore;
+        this.streamMetadataTasks = streamMetaTasks;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(final CreateReaderGroupEvent request) {
+        String scope = request.getScopeName();
+        String readerGroup = request.getRgName();
+        long requestId = request.getRequestId();
+        final RGOperationContext context = streamMetadataStore.createRGContext(scope, readerGroup);
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getVersionedReaderGroupState(scope, readerGroup,
+                true, context, executor)
+                .thenCompose(state -> {
+                    if (state.getObject().equals(ReaderGroupState.CREATING)) {
+                        String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
+                        return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
+                               .thenCompose(configRecord -> {
+                               if (!ReaderGroupConfig.StreamDataRetention.values()[configRecord.getObject().getRetentionTypeOrdinal()]
+                                   .equals(ReaderGroupConfig.StreamDataRetention.NONE)) {
+                                   // update Stream metadata tables, only if RG is a Subscriber
+                                   Iterator<String> streamIter = configRecord.getObject().getStartingStreamCuts().keySet().iterator();
+                                   return Futures.loop(() -> streamIter.hasNext(), () -> {
+                                          Stream stream = Stream.of(streamIter.next());
+                                          return streamMetadataStore.addSubscriber(stream.getScope(),
+                                                       stream.getStreamName(), scopedRGName, configRecord.getObject().getGeneration(),
+                                                  null, executor);
+                                      }, executor);
+                                   }
+                               return CompletableFuture.completedFuture(null);
+                               }).thenCompose(v ->
+                                  streamMetadataTasks.createRGStream(scope, NameUtils.getStreamForReaderGroup(readerGroup),
```

Comment:
i would prefer if we moved this call out of event processor.. the problem is it can lead to deadlocks.. one event processing waiting on another asynchronous processing and there may not be sufficient processing slots available to process the new one as the current request holds onto a processing slot. 
that is why i prefer that we create the RG stream and then submit the createRG event to be processed which simply manages the subscriptions. 

PS: i know that presently create stream is performed using task framework. 

an alternative here is to call `createStreamBody` however, that does not guarantee the mutual exclusion.

Reply:
also I am ok if you open a separate github issue for this and address it immediately after we close this PR.

Reply:
Created https://github.com/pravega/pravega/issues/5438